### ISSUE TITLE
strdup may cause segfault when std=c11 is used

### ISSUE THREAD

**Issue by jokester** [htacg/tidy-html5#636]
When `-std=c11` is used, call to `strdup` may return incorrect result and cause segfault. (It returned a 32bit pointer in my gcc/linux64 build, but works fine in clang/freebsd).

This can be avoided by using internal `tmbstrdup` instead, see [this (conceptual) diff](https://github.com/htacg/tidy-html5/compare/release/5.4...jokester:build-c11-5.4). Do you think this should be fixed?

**Comment by balthisar** [comment]
Can you do it without changing the compiler to c11, though? Tidy is meant to be buildable with c89, although I personally seem to require c89gnu.

**Comment by jokester** [comment]
Sure, I only mean to make it a little more standard complaint.

The fix I proposed should not depend on specific version of C, but I may require some help in not breaking things (not much experience in c/portable here).

BTW tidy may have some unrevealed requirement on C version? I could build last 5.4.0 release with `-std=gnu89` but not `-std=c89`

**Comment by geoffmcl** [comment]
@jokester thanks for the issue...

Seems to be at least two questions here -
 
 1. Use of `strdup` versus `tmbstrdup` in library
 2. GCC C standards, and portability
 
First, concerning the use of `strdup` in the **library**, I feel this is wrong. **ALL** memory allocations in the library should use **TidyAlloc**, ie `tmbstrdup`...

This is so that other apps using `libtidy` can supply their own memory allocator, especially including what to do in case of a `failure`, ie **panic**... When the library is part of a bigger app, it may not be appropriate to just `exit` that app, so they can add their own memory handler...

So my suggested patch for that, which includes replacing the `free` also, is -

```diff
diff --git a/src/language.c b/src/language.c
index 6ef97f5..4f758ea 100644
--- a/src/language.c
+++ b/src/language.c
@@ -335,7 +335,7 @@ tmbstr TY_(tidyNormalizedLocaleName)( ctmbstr locale )
     uint i;
     uint len;
     static char result[6] = "xx_yy";
-    tmbstr search = strdup(locale);
+    tmbstr search = TY_(tmbstrdup)(&TY_(g_default_allocator), locale);
     search = TY_(tmbstrtolower)(search);
     
     /* See if our string matches a Windows name. */
@@ -343,8 +343,8 @@ tmbstr TY_(tidyNormalizedLocaleName)( ctmbstr locale )
     {
         if ( strcmp( localeMappings[i].winName, search ) == 0 )
         {
-            free(search);
-            search = strdup(localeMappings[i].POSIXName);
+            TidyFree(&TY_(g_default_allocator), search);
+            search = TY_(tmbstrdup)(&TY_(g_default_allocator), localeMappings[i].POSIXName);
             break;
         }
     }
@@ -376,7 +376,7 @@ tmbstr TY_(tidyNormalizedLocaleName)( ctmbstr locale )
         }
     }
     
-    free( search );
+    TidyFree(&TY_(g_default_allocator), search);
     return result;
 }
 
```

Good catch...

As to the GCC standard to use, I will leave that to others, but it is important to understand that MS compilers may not exactly follow the **same** GCC standard...

And given that `tidy` is a relatively simple `C` project, it should be compileable by the oldest MS compiler possible, within reason... at this time at least MSVC10 (2010), which I understand did not, could not!, have `full` c11 (2011) compliance...

Not much experience with older versions of GCC... but have no problems with my Ubuntu 14.04, gcc 4.8.4, circa 2013...

Look forward to further feedback... thanks...

**Comment by jokester** [comment]
@geoffmcl thanks for pointing out the direction. Using a (maybe optional) custom allocator seems a more complete fix. I'll see what I can do with it.

Regarding the c version to use: if possible I would prefer to not break things, i.e. reduce inconsistency with c11, while keep existing versions work (do we have a known-to-work list on this?).

**Comment by balthisar** [comment]
There's not a formal known-work list, but I used Clang with gnu89 and GCC with gnu89 (straight C89 fails; I'll have to investigate this further), and MSVC 10 on Windows, where I usually catch my errors from not following strict C89 (e.g., `for ( int i = ... )` not opening new context in older C.

Looks like there are still a lot of definitions in `platform.c` for lots of compilers, but that doesn't indicate their flavor of C, of course.

**Comment by geoffmcl** [comment]
@jokester look forward to the replacement of `strdup/free` appropriately... thanks...

@balthisar just tried a MSVC 10 build, my first for a while, and it failed! Needed the following small `context` change -

```diff
diff --git a/src/config.c b/src/config.c
index 2a40f15..2f6bc16 100644
--- a/src/config.c
+++ b/src/config.c
@@ -581,8 +581,9 @@ static Bool subDeprecatedOption( TidyDocImpl* doc, ctmbstr oldName, ctmbstr oldV
             }
             else
             {
+                ctmbstr val;
                 TY_(SetOptionInt)( doc, newOptId, value );
-                ctmbstr val = tidyOptGetCurrPick( tdoc, newOptId );
+                val = tidyOptGetCurrPick( tdoc, newOptId );
                 TY_(Report)( doc, NULL, NULL, OPTION_REMOVED_APPLIED, oldName, newName, val );
             }
         }
```

Will try to get around to adding this MSVC 10 build more frequently... it is a good test...

I have tried several times before to add some flag to `gcc` to **show** up this, but have never succeeded! See discussion in #460, #453, and maybe others...

I was able to gcc compile tidy using `-std=c11 -D_GNU_SOURCE`... the main problem seems to be **uint** not being defined, without `-D_GNU_SOURCE`, but **uint** is used a **lot** in tidy source... some 750 times... as can be seen in earlier discussions, this **uint** problem is something I do not understand fully...

Note a `typedef unsigned int uint;` is done in `tidyplatform.h`, which is where it is defined in the Windows build... see [Lines 544-555](https://github.com/htacg/tidy-html5/blob/next/include/tidyplatform.h#L544)... and note the **interesting** comment that starts this block...

And more puzzling, I specifically modified the CMakeLists.txt in `tidy` to do this, but in another `C` project noted that it seems, under **certain circumstances**, cmake generates a `Makefile` that **includes** the flags `-std=c11 -D_GNU_SOURCE`... but then in another `C` project it did **not** add these flags??? What were the **different** circumstances???

Seems lots to explore, discuss, and **understand** here, **before** changing anything...

Anyway, look forward to the simpler `strdup` fix... thanks...

**Comment by jokester** [comment]
Thanks @geoffmcl . I'm working on a PR about replace ad-hoc `malloc/free`s to `TidyAlloc/TidyFree`. At least it should settle the initial problem of this issue.

As of the `uint`: I too don't get the intent of existing platform-specific includes: what is the "hack", and why a common `typedef unsigned int uint` is not used. Guess we can leave that to another issue.

**Comment by geoffmcl** [comment]
@jokester assume you meant `strdup/free`, in the **library**, of which there are only 2, shown above... thanks...

Of course, `malloc/free` are the default allocators in the **library**, if a user does not supply his/her own, so do exist, but **once** only... 

And `malloc/free` can be **freely** used outside the **library**, as in the console app, `tidy.c`... there is no problem here...

And I have solved my `flag` problems. Seems cmake caches any flags used, and I think it was just that I did not clean up a previous build, i.e. at least delete `CMakeCache.txt`, before conducting further gcc flag experiments...

Concerning `uint`: In my Ubuntu 64-bit system, I do find `typedef unsigned int uint;`, in `/usr/include/x86_64-linux-gnu/sys/types.h` but it is under an `#ifdef __USE_MISC` guard, so I guess adding `-D_GNU_SOURCE` adds this... but only needed if you also add `-std=c11`...

But as you point out, that should be a separate issue, if it is thought there is some problem to be addressed here, but I see none at the moment... certainly see no problem maintaining `/* Old compatibility names for C types. /*`, that is used some 750 times in **libtidy**...

Look forward to a PR if you get the time, or I can apply the above patch... thanks...
